builtInFunctions/setGuardian.go

```diff
@@ -0,0 +1,270 @@
+
+const (
+	GuardiansKeyIdentifier     = "guardians"
+	BuiltInFunctionSetGuardian = "SetGuardian"
+)
+
+type Guardian struct {
+	Address         []byte
+	ActivationEpoch uint32
+}
+
+type Guardians struct {
+	Data []*Guardian
+}
+
+// BlockChainEpochHook is a light-weight blockchain hook,
+// which can be queried to provide the current epoch
+type BlockChainEpochHook interface {
+	CurrentEpoch() uint32
+	IsInterfaceNil() bool
+}
+
+// SetGuardianArgs is a struct placeholder for all necessary args
+// to create a NewSetGuardianFunc
+type SetGuardianArgs struct {
+	Marshaller      marshal.Marshalizer
+	PubKeyConverter core.PubkeyConverter
+	EpochNotifier   vmcommon.EpochNotifier
+	BlockChainHook  BlockChainEpochHook
+
+	FuncGasCost              uint64
+	GuardianActivationEpochs uint32
+	SetGuardianEnableEpoch   uint32
+}
+
+type setGuardian struct {
+	*baseEnabled
+	marshaller      marshal.Marshalizer
+	blockchainHook  BlockChainEpochHook
+	pubKeyConverter core.PubkeyConverter
+
+	mutExecution             sync.RWMutex
+	guardianActivationEpochs uint32
+	funcGasCost              uint64
+	keyPrefix                []byte
+}
+
+// NewSetGuardianFunc will instantiate a new set guardian built-in function
+func NewSetGuardianFunc(args SetGuardianArgs) (*setGuardian, error) {
+	if check.IfNil(args.Marshaller) {
+		return nil, ErrNilMarshaller
+	}
+	if check.IfNil(args.BlockChainHook) {
+		return nil, ErrNilBlockChainHook
+	}
+	if check.IfNil(args.PubKeyConverter) {
+		return nil, ErrNilPubKeyConverter
+	}
+	if check.IfNil(args.EpochNotifier) {
+		return nil, ErrNilEpochNotifier
+	}
+
+	setGuardianFunc := &setGuardian{
+		funcGasCost:              args.FuncGasCost,
+		marshaller:               args.Marshaller,
+		blockchainHook:           args.BlockChainHook,
+		pubKeyConverter:          args.PubKeyConverter,
+		guardianActivationEpochs: args.GuardianActivationEpochs,
+		mutExecution:             sync.RWMutex{},
+		keyPrefix:                []byte(core.ElrondProtectedKeyPrefix + GuardiansKeyIdentifier),
+	}
+	setGuardianFunc.baseEnabled = &baseEnabled{
+		function:        BuiltInFunctionSetGuardian,
+		activationEpoch: args.SetGuardianEnableEpoch,
+		flagActivated:   atomic.Flag{},
+	}
+
+	logAccountFreezer.Debug("set guardian enable epoch", args.SetGuardianEnableEpoch)
+	args.EpochNotifier.RegisterNotifyHandler(setGuardianFunc)
+
+	return setGuardianFunc, nil
+}
+
+// ProcessBuiltinFunction will process the set guardian built-in function call
+// Currently, the following cases are treated:
+// Case 1. User does NOT have any guardian => set guardian
+// Case 2. User has ONE guardian pending => do not set new guardian, wait until first one is set
+// Case 3. User has ONE guardian enabled => set guardian
+// Case 4. User has TWO guardians. FIRST is enabled, SECOND is pending => does not set, wait until second one is set
+// Case 5. User has TWO guardians. FIRST is enabled, SECOND is enabled => replace oldest one + set new one as pending
```

Comment:
I think we have to re talk this design. we can have a different design with multi signature. We could add as many addresses as we want first, set a limit of needed signatures for any action. If there are enough signatures action is performed, otherwise not. This action can be a moveBalance, an esdtTrasnfer, a contract call, or adding and removing addresses.

Reply:
This is too hard coded - we need a more relaxed guardian functionality. And it should be independent of epochs, it needs to protect more than unstaking and unbonding. Maiar DEX farming positions are higher for many users than eGLD staked positions.

Reply:
As discussed, we will keep the current structure as it is for now